docs: add guide for rewrites (#948)
ory/docs: docs/keto/reference/ory-permission-language.mdx

---
title: The "Ory Permission Language" Specification

> ---

Reviewer: Can you please use the same frontmatter here as in the other document? 
Please remember to sentence case everything aside from proper names (Ory Permission Language)

Enforcing fine-grained permissions is a critical building block of mature technology solutions that protect privacy and identity

> in the information age. Several proprietary languages to represent permission already exist, namely the Most permissions are

Reviewer: ```suggestion
in the information age. Several proprietary languages used to represent permission already exist, namely the Most permissions are
```

Reviewer: There's something wrong with this sentence, something missing after "namely"

defined by developers who are likely familiar with Web technologies like JavaScript or Typescript. There is a need for a

> developer-friendly configuration language for permissions that has such a small learning curve that most developers can understand

Reviewer: ```suggestion
developer-friendly configuration language for permissions that has a small learning curve small enough so that most developers can understand
```

> and use it with close to no effort. We therefore chose to define our permissions configuration language as a subset of the most

Reviewer: ```suggestion
and use it with minimal effort. To fulfill this need, we defined the permissions configuration language as a subset of the most
```

| `=`      | `permits = {...}`                        | Assignment token                                                  |
| `,`      | `{x: 1, y: 2}`                           | Property separator                                                |
| `'`      | `'string'`                               | Single quoted string literal                                      |
| `"`      | `"string"`                               | Double quoted string literal                                      |
| `*`      | `import * from @ory/permission-language` | Import glob                                                       |
| `\|`     | `(User \| Group)[]`                      | Type union                                                        |

## Statements

### Type declaration

The top level of the configuration consists of a list of `class` declarations for each namespace. Each `class` consists of
relation declarations and permission declarations.

```ebnf
Config          = [ ClassDecl ] .
ClassDecl       = "class" identifier "implements" "Namespace" "{" ClassSpec "}" .
ClassSpec       = [ RelationDecls ] | [ PermissionDefns] .
```

> The following example declares the type _User_.

Reviewer: ```suggestion
The following example declares the type `User`.
```
For better legibility, please refrain from using italics formatting.

The `related` section of type declarations defines relations. Unlike regular TypeScript, `RelationName` must be a unique
identifier used as the relation strings in the tuples. The `TypeName` must be the name of another `type` that is defined above or
below (in TypeScript: a class that implements `Namespace`).

Type unions (`|`) can be used to denote that a relation can have subjects of multiple types, e.g.,
`viewers: (User | SubjectSet<Group, "members">)[]`, meaning that the subject of the "viewer" relation can be either a "User", or a
subject set "Group#members".

```ebnf
RelationDecls   = "related" "=" "{" { RelationName ":" ArrayType } "}" .
RelationName    = identifier .
ArrayType       = RelationType | ( "(" RelationType { "|" RelationType } ")" ) "[]" .
RelationType    = SubjectType | SubjectSetType .
SubjectType     = TypeName .
SubjectSetType  = "SubjectSet" "<" TypeName, string_lit ">" .
TypeName        = identifier .
```

Note that all relations are defined as array types `T[]` because there are naturally only many-to-many relations in Keto.

The following declares a type _Document_ with three relations: _owners_ and _viewers_, both of which have _users_ as subjects.

> Additionally, the relation _parent_ has type _Document_.

Reviewer: Please remove the italics and replace with `code` when necessary and suitable.

```ts
class User {}

class Document {
  related = {
    parents: Document[]
    owners: User[]
    viewers: User[]
  }
}
```

### Permission definition

Permissions are defined as functions within class declarations that take a parameter `ctx` of type `Context` and evaluate to a
boolean `true` or `false`.

The type annotations for `ctx` and the return value are optional.

```ebnf
PermissionDefns = "permits" "=" "{" Permission [ "," Permission ] "}" .
Permission      = PermissionSign "=>" PermissionBody .
PermissionSign  = PermissionName ":" "(" "ctx" [ ":" "Context" ] ")" [ ":" "boolean" ] .
PermissionName  = identifier .
```

The `ctx` object is a fixed parameter that contains the `subject` for which the permission check should be conducted:

```ts
ctx = { subject: "some_user_id" }
```

> The context will contain more fields in the future, e.g., the IP range or geolocation, the time of day, or the security level of

Reviewer: ```suggestion
The context will contain more fields in the future, for example, the IP range or geolocation, the time of day, or the security level of
```

---
id: ory-permission-language
title: The "Ory Permission Language" specification
sidebar_label: Ory Permission Language spec
---

Enforcing fine-grained permissions is a critical building block of mature technology solutions that protect privacy and identity

> in the information age. Several proprietary languages used to represent permission already exist, namely the Most permissions are
> defined by developers who are likely familiar with Web technologies like JavaScript or Typescript. There is a need for a

Reviewer 2: ```suggestion
in the information age. Several proprietary languages used to represent permission already exist, such as Rego or Casbin. Most permissions are
defined by developers who are likely familiar with Web technologies like JavaScript or Typescript. There is a need for a
```
